File: 1062.py
from sys import stdin

N, K = map(int, stdin.readline().split())

words = []
for _ in range(N):
    words += [stdin.readline()[4:-5]]

visited = [False for _ in range(26)]

visited[ord('a') - ord('a')] = True
visited[ord('c') - ord('a')] = True
visited[ord('i') - ord('a')] = True
visited[ord('n') - ord('a')] = True
visited[ord('t') - ord('a')] = True

res = 0

def learned(alphabet, num_learned):
    global visited
    global K
    global res
    
    if num_learned == K:
        cnt = 0
        
        for i in range(N):
            for j in range(len(words[i])):
                if not visited[ord(words[i][j]) - ord('a')]:
                    break
            else:
                cnt += 1
                
        res = max(res, cnt)
        
    else:
        for a in range(alphabet, 26):
            if not visited[a]:
                visited[a] = True
                learned(a, num_learned + 1)
                visited[a] = False
            
if K < 5:
    print(0)
elif K == 26:
    print(N)
else:
    K -= 5
    learned(0, 0)
    print(res)

File: 1062.cpp
#include <iostream>
#include <vector>
#include <string>
#include <algorithm>
using namespace std;

int res;
int N;
int K;
bool visited[26];
vector<string> words;

void learned(int alphabet, int num_learned) {
    if(num_learned == K) {
        int cnt = 0;
        bool flag = true;
        
        for(int i = 0; i < N; i++) {
            flag = true;
            for(int j = 0; j < words[i].length(); j++) {
                if (!(visited[words[i][j] - 'a'])) {
                    flag = false;
                    break;
                }
            }
            if (flag) {
                cnt++;
            }
        }
        
        res = max(res, cnt);
    }
    else {
        for(int i = alphabet; i < 26; i++) {
            if(!visited[i]) {
                visited[i] = true;
                learned(i, num_learned + 1);
                visited[i] = false;
            }
        }
    }
}

int main() {
    ios_base :: sync_with_stdio(false); 
    cin.tie(NULL); 
    cout.tie(NULL);
    string str;
    
    cin >> N >> K;
    words.resize(N);
    
    for(int i = 0; i < N; i++) {
        cin >> words[i];
        words[i] = words[i].substr(4, words[i].length() - 8);
    }
    
    visited['a' - 'a'] = true;
    visited['c' - 'a'] = true;
    visited['i' - 'a'] = true;
    visited['n' - 'a'] = true;
    visited['t' - 'a'] = true;
    
    if (K < 5) {
        cout << 0;
    }
    else if (K == 26) {
        cout << N;
    }
    else {
        K -= 5;
        learned(0, 0);
        cout << res;
    }
}